pkg/server/commands/list_objects.go

```diff
@@ -159,14 +160,15 @@ func (q *ListObjectsQuery) evaluate(
 			close(connectedObjectsResChan)
 		}()
 
-		limitedTupleReader := storagewrappers.NewBoundedConcurrencyTupleReader(q.Datastore, q.CheckConcurrencyLimit)
+		limitedTupleReader := storagewrappers.NewBoundedConcurrencyTupleReader(q.Datastore, q.MaxConcurrentReads)
 
 		checkResolver := graph.NewLocalChecker(
 			storage.NewCombinedTupleReader(limitedTupleReader, req.GetContextualTuples().GetTupleKeys()),
-			q.CheckConcurrencyLimit,
+			q.ResolveNodeBreadthLimit,
+			q.MaxConcurrentReads,
```

Comment:
This code isn't pretty: callers of `NewLocalChecker` shouldn't have to remember to create these storage wrappers.

In a follow-up PR I would do

```go
checkResolver := graph.NewLocalChecker(q.ds,
			q.ResolveNodeBreadthLimit,
			q.MaxConcurrentReads)
```

and then, when calling `checkResolver.Execute`, I do this

```go
c.ds = storage.NewCombinedTupleReader(storagewrappers.NewBoundedConcurrencyTupleReader(c.ds, c.MaxConcurrentReads), req.GetContextualTuples().GetTupleKeys()),
```

Reply:
Why is the `ds` bound to the `req` in your example? Decorating the incoming request with a deeply internal ref to a database client has code smells. Instead, if you need to decorate the datastore per request (e.g. for contextual tuples), then that's a perfect use for context. I know that's a controversial topic, but it is a good usage for context.

```go
ctx = WithDatastoreContext(storage.NewCombinedTupleReader(ds))

ds := DatastoreFromContext(ctx)
```

Otherwise I'd say the next best would be explicit propagation via parameter passing.

Reply:
> Why is the ds bound to the req in your example

because i don't want to overwrite `c.ds` on every request, since every request has different contextual tuples

> be explicit propagation via parameter passing.

okay yeah i prefer this :)